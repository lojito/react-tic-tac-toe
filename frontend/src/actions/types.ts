import {
  Categories,
  FirstPlayer,
  GameResult,
  ImageName,
  PlayingLevel,
  SquareLocation,
} from "../types";

export type CategoryAction = {
  type: "CHANGE_CATEGORY";
  category: Categories;
};

export type DisabledAction = {
  type: "CHANGE_DISABLED";
  disabled: boolean;
};

export type FirstAction = {
  type: "CHANGE_FIRST";
  first: FirstPlayer;
};

export type UserImageAction = {
  type: "CHANGE_USER_IMAGE";
  userImage: ImageName;
};

export type ComputerImageAction = {
  type: "CHANGE_COMPUTER_IMAGE";
  computerImage: ImageName;
};

export type PlayingLevelAction = {
  type: "CHANGE_PLAYING_LEVEL";
  level: PlayingLevel;
};

export type GameOverAction = {
  type: "CHANGE_GAME_OVER";
  over: boolean;
};

export type ResultAction = {
  type: "CHANGE_GAME_RESULT";
  result: GameResult;
};

export type WinnersSquaresAction = {
  type: "CHANGE_WINNERS_SQUARES";
  winners: SquareLocation[];
};

export type ActionUnknow = {
  type: "UNKNOWN";
};

export type Action =
  | CategoryAction
  | DisabledAction
  | FirstAction
  | UserImageAction
  | ComputerImageAction
  | PlayingLevelAction
  | GameOverAction
  | ResultAction
  | WinnersSquaresAction
  | ActionUnknow;
