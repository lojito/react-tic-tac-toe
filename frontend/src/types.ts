export type CategoryId = 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9;

export type CategoryHavanaLandmarks = {
  id: 0;
  name: "CATEGORY_HAVANA_LANDMARKS";
  folder: "habana";
};

export type CategoryMontrealLandmarks = {
  id: 1;
  name: "CATEGORY_MONTREAL_LANDMARKS";
  folder: "montreal";
};

export type CategoryVancouverLandmarks = {
  id: 2;
  name: "CATEGORY_VANCOUVER_LANDMARKS";
  folder: "vancouver";
};

export type CategorySpainLandmarks = {
  id: 3;
  name: "CATEGORY_SPAIN_LANDMARKS";
  folder: "spain";
};

export type CategoryGermanyLandmarks = {
  id: 4;
  name: "CATEGORY_GERMANY_LANDMARKS";
  folder: "germany";
};

export type CategorySoccerPlayers = {
  id: 5;
  name: "CATEGORY_SOCCER_PLAYERS";
  folder: "soccer";
};

export type CategoryFruitsAndVegetables = {
  id: 6;
  name: "CATEGORY_FRUITS_AND_VEGETABLES";
  folder: "fruits";
};

export type CategoryAnimals = {
  id: 7;
  name: "CATEGORY_ANIMALS";
  folder: "animals";
};

export type CategoryPuppies = {
  id: 8;
  name: "CATEGORY_PUPPIES";
  folder: "puppies";
};

export type CategorySeinfeld = {
  id: 9;
  name: "CATEGORY_SEINFELD";
  folder: "seinfeld";
};

export type Categories =
  | CategoryAnimals
  | CategoryFruitsAndVegetables
  | CategoryGermanyLandmarks
  | CategoryHavanaLandmarks
  | CategoryMontrealLandmarks
  | CategoryPuppies
  | CategorySeinfeld
  | CategorySoccerPlayers
  | CategorySpainLandmarks
  | CategoryVancouverLandmarks;

export type ImageFolder =
  | "habana"
  | "montreal"
  | "vancouver"
  | "spain"
  | "germany"
  | "soccer"
  | "fruits"
  | "animals"
  | "puppies"
  | "seinfeld";

export type ImageName =
  | 0
  | 1
  | 2
  | 3
  | 4
  | 5
  | 6
  | 7
  | 8
  | 9
  | 10
  | 11
  | 12
  | 13
  | 14
  | 15
  | 16
  | 17
  | 18
  | 19;

export type FirstPlayer = Player.User | Player.Computer;

export enum PlayingLevel {
  Easy,
  Smart,
}

export enum Player {
  User,
  Computer,
  Nobody,
}

export type GameBoard = [
  Player,
  Player,
  Player,
  Player,
  Player,
  Player,
  Player,
  Player,
  Player
];

export type SquareBoard = {
  disabled: boolean;
  id: string;
  imagePath: string;
  className: "win" | "";
};

export enum SquareLocation {
  TopLeft,
  TopCenter,
  TopRight,
  MiddleLeft,
  MiddleCenter,
  MiddleRight,
  BottomLeft,
  BottomCenter,
  BottomRight,
}

export enum GameResult {
  Draw,
  Lost,
  Started,
  Won,
}

export type DBGame = {
  _id?: string;
  board: GameBoard;
  categoryId: CategoryId;
  first: FirstPlayer;
  userImage: ImageName;
  computerImage: ImageName;
  level: PlayingLevel;
  result: GameResult;
  winners: SquareLocation[];
  user?: string;
  createdAt?: string;
  updatedAt?: string;
  __v?: number;
};
