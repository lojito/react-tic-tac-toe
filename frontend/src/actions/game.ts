import {
  Categories,
  FirstPlayer,
  GameResult,
  ImageName,
  PlayingLevel,
  SquareLocation,
} from "../types";

import {
  CategoryAction,
  ComputerImageAction,
  DisabledAction,
  FirstAction,
  GameOverAction,
  PlayingLevelAction,
  ResultAction,
  UserImageAction,
  WinnersSquaresAction,
} from "./types";

export const CHANGE_CATEGORY = "CHANGE_CATEGORY";
export const CHANGE_DISABLED = "CHANGE_DISABLED";
export const CHANGE_FIRST = "CHANGE_FIRST";
export const CHANGE_IMAGES = "CHANGE_IMAGES";
export const CHANGE_USER_IMAGE = "CHANGE_USER_IMAGE";
export const CHANGE_COMPUTER_IMAGE = "CHANGE_COMPUTER_IMAGE";
export const CHANGE_PLAYING_LEVEL = "CHANGE_PLAYING_LEVEL";
export const CHANGE_GAME_OVER = "CHANGE_GAME_OVER";
export const CHANGE_GAME_RESULT = "CHANGE_GAME_RESULT";
export const CHANGE_WINNERS_SQUARES = "CHANGE_WINNERS_SQUARES";

export const changeCategory = (category: Categories): CategoryAction => {
  return {
    type: CHANGE_CATEGORY,
    category,
  };
};

export const changeDisabled = (disabled: boolean): DisabledAction => {
  return {
    type: CHANGE_DISABLED,
    disabled,
  };
};

export const changeFirst = (first: FirstPlayer): FirstAction => {
  return {
    type: CHANGE_FIRST,
    first,
  };
};

export const changeUserImage = (userImage: ImageName): UserImageAction => {
  return {
    type: CHANGE_USER_IMAGE,
    userImage,
  };
};

export const changeComputerImage = (
  computerImage: ImageName
): ComputerImageAction => {
  return {
    type: CHANGE_COMPUTER_IMAGE,
    computerImage,
  };
};

export const changeLevel = (level: PlayingLevel): PlayingLevelAction => {
  return {
    type: CHANGE_PLAYING_LEVEL,
    level,
  };
};

export const changeOver = (over: boolean): GameOverAction => {
  return {
    type: CHANGE_GAME_OVER,
    over,
  };
};

export const changeResult = (result: GameResult): ResultAction => {
  return {
    type: CHANGE_GAME_RESULT,
    result,
  };
};

export const changeWinners = (
  winners: SquareLocation[]
): WinnersSquaresAction => {
  return {
    type: CHANGE_WINNERS_SQUARES,
    winners,
  };
};
