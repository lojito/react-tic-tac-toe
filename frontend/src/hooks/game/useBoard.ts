import { useCallback, useEffect, useRef, useState } from "react";
import Board from "../../classes/Board";
import Computer from "../../classes/Computer";
import {
  Categories,
  DBGame,
  FirstPlayer,
  GameResult,
  ImageName,
  Player,
  PlayingLevel,
  SquareLocation,
} from "../../types";

interface Props {
  disabled: boolean;
  first: FirstPlayer;
  level: PlayingLevel;
  userImage: ImageName;
  computerImage: ImageName;
  category: Categories;
}

function useBoard(
  {
    disabled,
    first,
    level,
    userImage,
    computerImage,
    category: { id: categoryId },
  }: Props,
  onGameOver: (game: DBGame) => void
) {
  const board = useRef(Board);
  const computer = useRef(Computer);
  const gameOver = useRef(false);
  const player = useRef(Player.Nobody);
  const result = useRef(GameResult.Started);
  const winners = useRef<SquareLocation[]>([]);
  board.current.first = first;
  board.current.level = level;
  const [gameBoard, setGameBoard] = useState(board.current.getBoard());

  const playGame = useCallback((square?: SquareLocation) => {
    if (player.current === Player.Computer) {
      computer.current.play();
    } else {
      board.current.place(Player.User, square!);
    }
    setGameBoard(board.current.getBoard());

    winners.current = board.current.isAWinner(player.current);
    if (winners.current.length) {
      gameOver.current = true;
      result.current =
        player.current === Player.Computer ? GameResult.Lost : GameResult.Won;
    } else if (board.current.isFull) {
      gameOver.current = true;
      result.current = GameResult.Draw;
    } else {
      player.current =
        player.current === Player.Computer ? Player.User : Player.Computer;
      if (player.current === Player.Computer) {
        playGame();
      }
    }
  }, []);

  const resetGame = useCallback(() => {
    board.current.reset();
    gameOver.current = false;
    player.current = Player.Nobody;
    result.current = GameResult.Started;
    winners.current = [];
    setGameBoard(board.current.getBoard());
  }, []);

  useEffect(() => {
    if (!disabled && !gameOver.current) {
      resetGame();
    }
  }, [disabled, resetGame]);

  useEffect(() => {
    if (disabled === false) {
      if (first === Player.Computer) {
        player.current = Player.Computer;
      } else {
        player.current = Player.User;
      }
    } else {
      if (player.current === Player.Computer) {
        playGame();
      }
    }
  }, [disabled, first, playGame]);

  useEffect(() => {
    if (gameOver.current) {
      onGameOver({
        board: gameBoard,
        categoryId,
        first,
        userImage: userImage,
        computerImage: computerImage,
        level,
        result: result.current,
        winners: winners.current,
      });
      gameOver.current = false;
    }
  });

  return {
    playGame,
    gameBoard,
  };
}

export default useBoard;
