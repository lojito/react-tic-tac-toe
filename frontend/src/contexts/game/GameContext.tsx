import React, { createContext } from "react";
import { Action } from "../../actions/types";
import {
  Categories,
  FirstPlayer,
  GameResult,
  ImageName,
  PlayingLevel,
  SquareLocation,
} from "../../types";

export type State = {
  category: Categories;
  userImage: ImageName;
  computerImage: ImageName;
  first: FirstPlayer;
  level: PlayingLevel;
  disabled: boolean;
  over: boolean;
  result: GameResult;
  winners: SquareLocation[];
};

export const GameContext = createContext<{
  game: State;
  dispatch: React.Dispatch<Action>;
}>({ game: {} as State, dispatch: {} as React.Dispatch<Action> });
