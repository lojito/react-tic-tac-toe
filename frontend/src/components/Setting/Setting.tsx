import React, { FC, memo, useContext } from "react";
import { Action, FirstAction, PlayingLevelAction } from "../../actions/types";
import { GameContext } from "../../contexts/game/GameContext";
import { Player, PlayingLevel } from "../../types";
import "./Setting.scss";

interface Option {
  value: Player | PlayingLevel;
  text: string;
}

interface SettingProps {
  tooltip: string;
  label: string;
  defaultValue: Player | PlayingLevel;
  changeAction: (setting: number) => FirstAction | PlayingLevelAction;
  testid: string;
  options: Option[];
}

interface Props extends SettingProps {
  disabled: boolean;
  dispatch: React.Dispatch<Action>;
}

const MemoizedSetting: FC<Props> = memo(
  ({
    disabled,
    dispatch,
    tooltip,
    label,
    defaultValue,
    testid,
    changeAction,
    options,
  }) => {
    const optionsJSX = options.map(({ value, text }: Option) => {
      return (
        <option key={text} value={value}>
          {text}
        </option>
      );
    });
    return (
      <div className="setting" title={tooltip} data-testid={testid}>
        <label htmlFor="setting">{label}?</label>
        <select
          defaultValue={defaultValue}
          id="setting"
          onChange={(e) => {
            dispatch(changeAction(+e.target.value));
          }}
          disabled={disabled}
        >
          {optionsJSX}
        </select>
      </div>
    );
  }
);

const Setting: FC<SettingProps> = ({
  tooltip,
  label,
  defaultValue,
  changeAction,
  testid,
  options,
}) => {
  const { game, dispatch } = useContext(GameContext);

  return (
    <MemoizedSetting
      disabled={game.disabled}
      dispatch={dispatch}
      tooltip={tooltip}
      label={label}
      defaultValue={defaultValue}
      testid={testid}
      changeAction={changeAction}
      options={options}
    />
  );
};

export default Setting;
