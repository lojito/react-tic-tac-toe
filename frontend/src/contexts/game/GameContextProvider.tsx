import React, { FC, useMemo, useReducer } from "react";
import gameReducer from "../../reducers/gameReducer";
import { GameContext } from "./GameContext";
import initialState from "./GameInitialState";

const GameContextProvider: FC = ({ children }) => {
  const [game, dispatch] = useReducer(gameReducer, initialState);
  const value = useMemo(() => {
    return { game, dispatch };
  }, [game]);

  return <GameContext.Provider value={value}>{children}</GameContext.Provider>;
};

export default GameContextProvider;
