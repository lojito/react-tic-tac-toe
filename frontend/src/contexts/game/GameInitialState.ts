import { State } from "../../contexts/game/GameContext";
import categories from "../../data/json/categories.json";
import {
  Categories,
  GameResult,
  ImageName,
  Player,
  PlayingLevel,
} from "../../types";

const NUMBER_OF_CATEGORIES = 10;
const category = categories[
  Math.floor(Math.random() * NUMBER_OF_CATEGORIES)
] as Categories;

const IMAGES_PER_CATEGORY = 20;
const userImage = Math.floor(Math.random() * IMAGES_PER_CATEGORY) as ImageName;
const computerImage = Math.floor(
  Math.random() * IMAGES_PER_CATEGORY
) as ImageName;

const initialState: State = {
  category,
  userImage,
  computerImage,
  first: Player.User,
  level: PlayingLevel.Easy,
  disabled: false,
  over: false,
  result: GameResult.Started,
  winners: [],
};

export default initialState;
