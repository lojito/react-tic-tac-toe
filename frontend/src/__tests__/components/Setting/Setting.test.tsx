import React from "react";
import "@testing-library/jest-dom/extend-expect";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import Setting from "../../../components/Setting/Setting";
import { GameContext, State } from "../../../contexts/game/GameContext";
import { PlayingLevel as Level, Player } from "../../../types";

describe("Setting", () => {
  const mockDictionary = {
    PLAYING_LEVEL_LABEL: "Level",
    PLAYING_LEVEL_EASY: "Easy",
    PLAYING_LEVEL_SMART: "Smart",
    PLAYING_LEVEL_TOOLTIP:
      "How much thinking the computer does before playing.",
    START_GAME_LABEL: "Who plays first",
    START_GAME_TOOLTIP: "Who starts playing the game: You or the computer?",
    START_GAME_ANSWER_YES: "Yes",
    START_GAME_ANSWER_NO: "No",
  };

  let tooltip: string;
  let label: string;
  let testid: string;
  let option1Text: string;
  let option1Value: Level | Player;
  let option2Text: string;
  let option2Value: Level | Player;

  interface Props {
    disabled?: boolean;
    renderPlayingLevel?: boolean;
  }

  const renderSetting = ({
    disabled = false,
    renderPlayingLevel = true,
  }: Props = {}) => {
    if (renderPlayingLevel) {
      tooltip = mockDictionary.PLAYING_LEVEL_TOOLTIP!;
      label = mockDictionary.PLAYING_LEVEL_LABEL!;
      testid = "level";
      option1Text = mockDictionary.PLAYING_LEVEL_EASY!;
      option1Value = Level.Easy;
      option2Text = mockDictionary.PLAYING_LEVEL_SMART!;
      option2Value = Level.Smart;
    } else {
      tooltip = mockDictionary.START_GAME_TOOLTIP!;
      label = mockDictionary.START_GAME_LABEL!;
      testid = "startgame";
      option1Text = mockDictionary.START_GAME_ANSWER_YES!;
      option1Value = Player.User;
      option2Text = mockDictionary.START_GAME_ANSWER_NO!;
      option2Value = Player.Computer;
    }

    const game = {
      level: option1Value,
      disabled,
    } as State;

    const dispatch = jest.fn();
    const changeAction = jest.fn();

    const utils = render(
      <GameContext.Provider value={{ game, dispatch }}>
        <Setting
          tooltip={tooltip}
          label={label}
          defaultValue={option1Value}
          changeAction={changeAction}
          testid={testid}
          options={[
            {
              value: option1Value,
              text: option1Text,
            },
            {
              value: option2Value,
              text: option2Text,
            },
          ]}
        />
      </GameContext.Provider>
    );

    const divWrapper = screen.getByTestId(testid);

    const select = screen.getByRole("combobox");

    const options = screen.getAllByRole("option");

    const option1 = screen.getByRole("option", {
      name: option1Text,
    }) as HTMLOptionElement;

    const option2 = screen.getByRole("option", {
      name: option2Text,
    }) as HTMLOptionElement;

    return {
      ...utils,
      mockDictionary,
      dispatch,
      divWrapper,
      select,
      options,
      option1,
      option1Value,
      option2,
      option2Value,
    };
  };

  it("should render correctly", () => {
    const { asFragment } = renderSetting();

    expect(asFragment()).toMatchSnapshot();
  });

  it("should render correctly as a StartGame component", () => {
    const { asFragment } = renderSetting({ renderPlayingLevel: false });

    expect(asFragment()).toMatchSnapshot();
  });

  it("should display the dropdown", () => {
    const { select } = renderSetting();

    expect(select).toBeInTheDocument();
  });

  it("should apply the 'setting' class", () => {
    const { divWrapper } = renderSetting();

    expect(divWrapper).toHaveClass("setting");
  });

  it("should correctly set default option", () => {
    const { option1 } = renderSetting();

    expect(option1.selected).toBeTruthy();
  });

  it("should display the correct number of options", () => {
    const { options } = renderSetting();

    expect(options).toHaveLength(2);
  });

  it("should allow user to change the dropdown default option and call the dispatch function when disabled is false", async () => {
    const user = userEvent.setup();
    const { dispatch, select, option1, option2, option1Value, option2Value } =
      renderSetting();

    expect(select).toHaveValue(option1Value + "");
    expect(option1.selected).toBeTruthy();
    expect(option2.selected).toBeFalsy();

    await user.selectOptions(select, option2);

    expect(select).toHaveValue(option2Value + "");
    expect(option1.selected).toBeFalsy();
    expect(option2.selected).toBeTruthy();
    expect(dispatch).toBeCalled();
  });

  it("should not allow the user to change the dropdown default option nor call the dispatch function when disabled is true", () => {
    const user = userEvent.setup();
    const { dispatch, select, option2, option1Value } = renderSetting({
      disabled: true,
    });

    expect(select).toHaveValue(option1Value + "");

    user.selectOptions(select, option2);

    expect(select).toHaveValue(option1Value + "");
    expect(dispatch).not.toBeCalled();
  });
});
