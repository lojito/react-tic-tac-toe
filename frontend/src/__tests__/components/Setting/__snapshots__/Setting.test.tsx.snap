// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Setting should render correctly 1`] = `
<DocumentFragment>
  <div
    class="setting"
    data-testid="level"
    title="How much thinking the computer does before playing."
  >
    <label
      for="setting"
    >
      Level?
    </label>
    <select
      id="setting"
    >
      <option
        selected=""
        value="0"
      >
        Easy
      </option>
      <option
        value="1"
      >
        Smart
      </option>
    </select>
  </div>
</DocumentFragment>
`;

exports[`Setting should render correctly as a StartGame component 1`] = `
<DocumentFragment>
  <div
    class="setting"
    data-testid="startgame"
    title="Who starts playing the game: You or the computer?"
  >
    <label
      for="setting"
    >
      Who plays first?
    </label>
    <select
      id="setting"
    >
      <option
        selected=""
        value="0"
      >
        Yes
      </option>
      <option
        value="1"
      >
        No
      </option>
    </select>
  </div>
</DocumentFragment>
`;
